src/Parsers/PRQL/ParserPRQLQuery.h

```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include <Parsers/IParserBase.h>
+
+namespace DB
+{
+class ParserPRQLQuery final : public IParserBase
+{
+private:
+    // These fields are not used when PRQL is disabled at build time
+    [[maybe_unused]] size_t max_query_size;
```

Comment:
Should the whole class not be declared if clickhouse compiled w/o PRQL? Then this will go away

Reply:
I wanted to do the same. But in order to do that we should prevent the users setting the `dialect` settings to `prql`. Right now the settings are handled by macros:
https://github.com/ClickHouse/ClickHouse/blob/9953426cb08948857e09998d2493344bbddd76a9/src/Core/SettingsEnums.h#L203-L211
and
https://github.com/ClickHouse/ClickHouse/blob/9953426cb08948857e09998d2493344bbddd76a9/src/Core/SettingsEnums.cpp#L166-L169

So I had to "eliminate" the `prql` value from the enum. In the header that is doable:
```cpp
 enum class Dialect 
 { 
     clickhouse, 
     kusto, 
     kusto_auto,
#if USE_PRQL
     prql, 
#endif
 }; 
```
However in the source file in cannot work like that, more like I have to double the definition:
```cpp
#if USE_PRQL
 IMPLEMENT_SETTING_ENUM(Dialect, ErrorCodes::BAD_ARGUMENTS, 
     {{"clickhouse", Dialect::clickhouse}, 
      {"kusto", Dialect::kusto}, 
      {"prql", Dialect::prql}}) 
#elif
 IMPLEMENT_SETTING_ENUM(Dialect, ErrorCodes::BAD_ARGUMENTS, 
     {{"clickhouse", Dialect::clickhouse}, 
      {"kusto", Dialect::kusto}, 
      {"prql", Dialect::prql}}) 
#endif
```

I thought handling the dialect this way is better, because:
1. it doesn't require duplicating small chunk of code
2. easier to handle dialects without `#if`s on both client and server side

However I don't have a strong preference, if you think this is the better approach I am happy to make the necessary changes.

Reply:
I thought about moving this check into `executeQueryImpl`, then you can define the ParserPRQLQuery class only if if PRQL had been enabled.

Reply:
But then how would you prevent the users from setting the dialect to `PRQL`? Because we can do something like this:
```cpp
if (settings.dialect == Dialect::kusto && !internal)
        {
            ParserKQLStatement parser(end, settings.allow_settings_after_format_in_insert);

            /// TODO: parser should fail early when max_query_size limit is reached.
            ast = parseQuery(parser, begin, end, "", max_query_size, settings.max_parser_depth);
        }
#if USE_PRQL
        else if (settings.dialect == Dialect::prql && !internal)
        {
            ParserPRQLQuery parser(max_query_size, settings.max_parser_depth);
            ast = parseQuery(parser, begin, end, "", max_query_size, settings.max_parser_depth);
        }
#endif
        else
        {
            ParserQuery parser(end, settings.allow_settings_after_format_in_insert);
            /// TODO: parser should fail early when max_query_size limit is reached.
            ast = parseQuery(parser, begin, end, "", max_query_size, settings.max_parser_depth);
        }
```

But this can be confusing, because if the user executes `SET dialect = 'prql'`, it will succeed but actually it will result in using the `clickhouse` dialect. I think this is at least confusing, but maybe even wrong behavior.

If you meant something else that I don't see, please let me know. I am up for making this right, but I don't see how we can do this really nicely.

Reply:
Sorry that it hadn't been clear, but I meant move that exception from the PRQL to make the parser code cleaner:

```c
else if (settings.dialect == Dialect::prql && !internal)
{
#if USE_PRQL
    ParserPRQLQuery parser(max_query_size, settings.max_parser_depth);
    ast = parseQuery(parser, begin, end, "", max_query_size, settings.max_parser_depth);
#else
    throw Exception(
        ErrorCodes::SUPPORT_IS_DISABLED, "PRQL is not available. Rust code or PRQL itself may be disabled. Use another dialect!");
#endif
}
```

Reply:
But then I think it is not possible to change the dialect back to `clickhouse`, so in case we don't support PRQL setting the dialect to `prql` basically makes your session unusable. Is that something we really want?

Reply:
Additionally we could try to parse the set query there, but then we have to do the same thing in client and server also.

Reply:
>But then I think it is not possible to change the dialect back to clickhouse, so in case we don't support PRQL setting the dialect to prql basically makes your session unusable. Is that something we really want?

Oh, I see, my bad - you are right, then let's keep it this way.

Also maybe it worth to put a comment into the parser code why it cannot be written different way, just a brief summary of this thread.